### DISCUSSION TITLE
OpenTelemetry Jaeger endpoint

### DISCUSSION THREAD

**Discussion by PandeoF1** [keycloak/keycloak#28505] [Q&A] [answered]
Hi, i want to have some trace of keycloak using OpenTelemetry.
I have see that there is a page about [this](https://www.keycloak.org/keycloak-benchmark/kubernetes-guide/latest/util/otel), but the problem is that it say "Contrary to other setups, this is not using an OpenTelemetry collector, but instead exposes the metrics directly via Prometheus and sends traces directly to Jaeger."

If the agent sent directly the data to jaeger, how can i set the endpoint to hit.
Because using  the ENV var `OTEL_EXPORTER_OTLP_ENDPOINT` it doesn't work.

**Comment by codespearhead** [comment]
Try [setting environment variable](https://www.keycloak.org/server/configuration-metrics) `KC_METRICS_ENABLED=true` and hitting `/metrics`.

**Reply by PandeoF1** [reply]
For the metrics it work (exactly like how prometheus exporter work) but i want to have some trace from keycloak (i imagine that is possible because OTEL is made to generate trace logs and metrics)

**Reply by codespearhead** [reply]
I see. Have you tried [this](https://github.com/jangaraj/keycloak-with-opentelemetry/blob/548d233c46c1b9d7c2aa7f79d5abe0ecd05afe22/docker-compose.yaml#L22).

**Reply by PandeoF1** [reply]
I have tried to use KC_OTEL=true but i didn't tried with JAVA_OPTS_APPEND.
I will try that tomorrow !

**Reply by PandeoF1** [reply]
Update: It will take more time to adapt our images of keycloak so i will close this and reopen it when it will be tested

**Reply by omarfakhreddine** [reply]
@PandeoF1 Hello, I'm looking into this now, have you tried this yet?

**Reply by PandeoF1** [reply]
> @PandeoF1 Hello, I'm looking into this now, have you tried this yet?

Hey !

I have directly set up a Java agent to export everything from Keycloak because nothing else was working. 

Currently, I'm using the [OpenTelemetry Java Instrumentation Agent](https://github.com/open-telemetry/opentelemetry-java-instrumentation) to export traces, metrics, and logs from Keycloak, and it works pretty well.

My Keycloak is deployed in a k3s cluster, so I use the [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) with the annotation `instrumentation.opentelemetry.io/inject-java: "true"` (it automatically deploys the agent and configures everything). I send everything to my [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector-contrib). This collector sends the traces to the Jaeger collector, the logs directly to my OpenSearch cluster, and the metrics to Prometheus.

So yes it work but inside a kubernetes it's worth to use the operator of opentelemetry ^^

**Comment by paulomorgado** [comment, marked as answer]
I followed the instructions on https://www.keycloak.org/server/tracing and am launching the container with:

`start-dev --import-realm --hostname-strict=false --http-enabled=true --verbose --tracing-enabled=true --features=opentelemetry`

and:

`-e OTEL_EXPORTER_OTLP_ENDPOINT=http://host.containers.internal:19270 -e KC_OTEL=true ...`

not even setting `QUARKUS_OTEL_EXPORTER_OTLP_ENDPOINT` worked.

**Reply by mabartos** [reply]
Yes, this is a correct approach. AFAIK, instead of setting these env vars, you should be able to achieve the same by specifying `--tracing-endpoint=http://host.containers.internal:19270`.

Note: Keycloak 26.1 release will bring more capabilities for tracing, so you can check it out when it's live.

**Reply by paulomorgado** [reply]
What's the correct approach? `OTEL_EXPORTER_OTLP_ENDPOINT` or `QUARKUS_OTEL_EXPORTER_OTLP_ENDPOINT`?

No matter what I set, I always get:

```
2025-01-13T16:01:50 2025-01-13 16:01:50,063 traceId=, parentId=, spanId=, sampled= WARNING [io.quarkus.opentelemetry.runtime.exporter.otlp.sender.VertxGrpcSender] (vert.x-eventloop-thread-1) Failed to export TraceRequestMarshalers. The request could not be executed. Full error message: Connection refused: localhost/127.0.0.1:4317
2025-01-13T16:01:50 2025-01-13 16:01:50,066 traceId=, parentId=, spanId=, sampled= WARNING [io.opentelemetry.exporter.internal.grpc.GrpcExporter] (vert.x-eventloop-thread-1) Failed to export spans. Server responded with gRPC status code 2. Error message: Failed to export TraceRequestMarshalers. The request could not be executed. Full error message: Connection refused: localhost/127.0.0.1:4317
2025-01-13T16:01:50 
2025-01-13T16:01:50 UPDATE SUMMARY
2025-01-13T16:01:50 Run:                        148
2025-01-13T16:01:50 Previously run:               0
2025-01-13T16:01:50 Filtered out:                 0
2025-01-13T16:01:50 -------------------------------
2025-01-13T16:01:50 Total change sets:          148
2025-01-13T16:01:50 
2025-01-13T16:01:50 2025-01-13 16:01:50,132 traceId=, parentId=, spanId=, sampled= WARN  [io.agroal.pool] (main) Datasource '<default>': JDBC resources leaked: 1 ResultSet(s) and 0 Statement(s)
2025-01-13T16:01:50 2025-01-13 16:01:50,497 traceId=, parentId=, spanId=, sampled= INFO  [org.keycloak.connections.infinispan.DefaultInfinispanConnectionProviderFactory] (main) Node name: node_586907, Site name: null
2025-01-13T16:01:50 2025-01-13 16:01:50,640 traceId=, parentId=, spanId=, sampled= WARNING [io.quarkus.opentelemetry.runtime.exporter.otlp.sender.VertxGrpcSender] (vert.x-eventloop-thread-1) Failed to export TraceRequestMarshalers. The request could not be executed. Full error message: Connection refused: localhost/127.0.0.1:4317
```

Is there a preview image I can use for 26.1?

**Reply by paulomorgado** [reply]
@mabartos, do you have a working example of using keycloak's container image with OTEL?

**Reply by PandeoF1** [reply]
Hey @paulomorgado , on my side to get traces and metrics from Keycloak i use opentelemetry auto inject features with the annotations on my kube deployment :
        instrumentation.opentelemetry.io/inject-java: 'true'
It automatically add the java instrumentation agent to my pods.
If you are not under kubernetes you can just add this environments variables :
JAVA_TOOL_OPTIONS= -javaagent:/otel-auto-instrumentation-java/javaagent.jar'
(Download the java agent at the correct location on your computer)

The other env var that you will need are :
OTEL_SERVICE_NAME=keycloak
OTEL_EXPORTER_OTLP_ENDPOINT=http://your-collector:4318

**Reply by paulomorgado** [reply]
@PandeoF1, I don't think 26.0.8 comes with that:

```
2025-01-15T10:04:00
 Picked up JAVA_TOOL_OPTIONS: -javaagent:/otel-auto-instrumentation-java/javaagent.jar
2025-01-15T10:04:00
 Error opening zip file or JAR manifest missing : /otel-auto-instrumentation-java/javaagent.jar
```

**Reply by PandeoF1** [reply]
@paulomorgado You have this error because your image/server doesn't contains the javaagent of opentelemetry.
If you are using the keycloak docker image try to mount a volume with the java agent inside ^^
https://github.com/open-telemetry/opentelemetry-java-instrumentation/releases/tag/v2.11.0

**Reply by mabartos** [reply]
Keycloak 26.0.x, and 26.1.x includes Quarkus OpenTelemetry extension which contains all the necessary stuff for emitting traces (micrometer used for metrics).

The only missing part in your setup is the endpoint is not propagated.

Please, set the `KC_TRACING_ENDPOINT` env var and you should be set. 

The env var is mapped to `quarkus.otel.exporter.otlp.traces.endpoint` Quarkus property -> so it should be `QUARKUS_OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` env var.

**Reply by paulomorgado** [reply]
Something like this?

**Container:**

|Name|Value|
|-|-|
|Container image|quay.io/keycloak/keycloak:26.1.0|
|Container arguments|[ "/opt/keycloak/bin/kc.sh", "start-dev", "--import-realm", "--hostname-strict=false", "--http-enabled=true", "--verbose", "--tracing-enabled=true", "--features=opentelemetry" ]|
|Container ports|[ 8080, 9000 ]|

**Environment variables:**

|Name|Value|
|-|-|
|KC_HEALTH_ENABLED|true|
|KC_OTEL|true|
|KC_PROXY|edge|
|KC_TRACING_ENDPOINT|http://host.containers.internal:19270/v1/traces|
|KEYCLOAK_ADMIN|admin|
|KEYCLOAK_ADMIN_PASSWORD|admin|
|KEYCLOAK_LOGLEVEL|DEBUG|
|OTEL_BLRP_SCHEDULE_DELAY|1000|
|OTEL_BSP_SCHEDULE_DELAY|1000|
|OTEL_EXPORTER_OTLP_ENDPOINT|http://host.containers.internal:19270|
|OTEL_EXPORTER_OTLP_PROTOCOL|grpc|
|OTEL_METRICS_EXEMPLAR_FILTER|trace_based|
|OTEL_METRIC_EXPORT_INTERVAL|1000|
|OTEL_RESOURCE_ATTRIBUTES|service.instance.id=szfwsayw|
|OTEL_SERVICE_NAME|keycloak|
|OTEL_TRACES_SAMPLER|always_on|
|QUARKUS_HTTP_ACCESS_LOG_ENABLED|true|
|QUARKUS_HTTP_ACCESS_LOG_PATTERN|Request: %r %s %b %D %h %u %rn %t "%{i,Referer}" "%{i,User-Agent}" %{i,ALL_REQUEST_HEADERS} %{i,body} %{o,ALL_RESPONSE_HEADERS} %{o,body}|
